From 590b51b2cfeb2bfed9d9698f3447509d550c9004 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20Hr=C5=AFza?= <martin.hruza@braiins.cz>
Date: Tue, 8 Apr 2025 14:31:28 +0200
Subject: [PATCH] Adjust gain before scanning

- This patch temporarily increases the rx gain before scanning and reset
  it back to the last recorded value once the scanning finishes.
- Ref: https://github.com/lwfinger/rtw88/issues/337
---
 main.c |  3 +++
 phy.c  | 15 +++++++++++++++
 phy.h  |  2 ++
 3 files changed, 20 insertions(+)

diff --git a/main.c b/main.c
index 94c304c..cdfffe7 100644
--- a/main.c
+++ b/main.c
@@ -1571,6 +1571,8 @@ void rtw_core_scan_start(struct rtw_dev *rtwdev, struct rtw_vif *rtwvif,
 
 	set_bit(RTW_FLAG_DIG_DISABLE, rtwdev->flags);
 	set_bit(RTW_FLAG_SCANNING, rtwdev->flags);
+
+	rtw_phy_dig_set_max_coverage(rtwdev);
 }
 
 void rtw_core_scan_complete(struct rtw_dev *rtwdev, struct ieee80211_vif *vif,
@@ -1582,6 +1584,7 @@ void rtw_core_scan_complete(struct rtw_dev *rtwdev, struct ieee80211_vif *vif,
 	if (!rtwvif)
 		return;
 
+	rtw_phy_dig_reset(rtwdev);
 	clear_bit(RTW_FLAG_SCANNING, rtwdev->flags);
 	clear_bit(RTW_FLAG_DIG_DISABLE, rtwdev->flags);
 
diff --git a/phy.c b/phy.c
index 9103cb6..53c9483 100644
--- a/phy.c
+++ b/phy.c
@@ -372,6 +372,21 @@ static void rtw_phy_statistics(struct rtw_dev *rtwdev)
 #define DIG_CVRG_MIN				0x1c
 #define DIG_RSSI_GAIN_OFFSET			15
 
+void rtw_phy_dig_set_max_coverage(struct rtw_dev *rtwdev) {
+	rtw_dbg(rtwdev, RTW_DBG_PHY, "Setting IGI to max coverage\n");
+	rtw_phy_dig_write(rtwdev, DIG_CVRG_MAX);
+}
+
+void rtw_phy_dig_reset(struct rtw_dev *rtwdev) {
+	struct rtw_dm_info *dm_info = &rtwdev->dm_info;
+	u8 last_igi;
+
+	last_igi = dm_info->igi_history[0];
+	rtw_dbg(rtwdev, RTW_DBG_PHY, "Resetting IGI=%u\n", last_igi);
+
+	rtw_phy_dig_write(rtwdev, last_igi);
+}
+
 static bool
 rtw_phy_dig_check_damping(struct rtw_dm_info *dm_info)
 {
diff --git a/phy.h b/phy.h
index c9e6b86..8449936 100644
--- a/phy.h
+++ b/phy.h
@@ -146,6 +146,8 @@ static inline int rtw_check_supported_rfe(struct rtw_dev *rtwdev)
 }
 
 void rtw_phy_dig_write(struct rtw_dev *rtwdev, u8 igi);
+void rtw_phy_dig_reset(struct rtw_dev *rtwdev);
+void rtw_phy_dig_set_max_coverage(struct rtw_dev *rtwdev);
 
 struct rtw_power_params {
 	u8 pwr_base;
-- 
2.49.0

